gs_quant/api/fred/fredapi.py

```diff
@@ -0,0 +1,305 @@
+"""
+Copyright 2019 Goldman Sachs.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+
+//Portions copyright Maverick Lin. Licensed under Apache 2.0 license
+"""
+
+import os
+import sys
+from typing import Iterable, List, Optional, Tuple, Union
+
+import json
+import pandas as pd
+import datetime as dt
+
+import requests
+from requests.utils import requote_uri
+from requests.exceptions import HTTPError
+from urllib.parse import urlencode
+
+"""
+Fred Data API that provides functions to query the Fred dataset. 
+You need to specify a valid API key by passing in the string via api_key. 
+You can sign up for a free API key on the Fred website at http://research.stlouisfed.org/fred2/
+"""
+
+class FredDataApi(object):
+    earliest_realtime_start = '1776-07-04'
+    latest_realtime_end = '9999-12-31'
+    root_url = 'https://api.stlouisfed.org/fred'
+
+    def __init__(self, api_key=None):
+        if api_key is not None:
+            self.api_key = api_key
+        else:
+            import textwrap
```

Comment:
Please import at module-level. Scoped imports should only be used to avoid import circularities